### ISSUE TITLE
key-duplicates are not caught when using anchors

### ISSUE THREAD

**Issue by MarkusTeufelberger** [adrienverge/yamllint#175]
Consider the following example:

```
---
.foo: &foo
  my_key: value1

bar:
  <<: *foo
  my_key: value2
```

What should be the value of `bar.my_key`?

Note that this gives a `key-duplicates` error:

```
---
bar:
  my_key: value1
  my_key: value2
```

**Comment by Jongy** [comment]
I encountered this as well. Any plans to detect it in yamllint?

**Comment by perlpunk** [comment]
Merging/overriding keys is the whole point of the [merge key feature](https://yaml.org/type/merge.html) in YAML 1.1.

So your example is valid, the result would be:
```
---
.foo:
  my_key: value1
bar:
  my_key: value2
```
There is nothing to do here for yamllint. I suggest to close this.

**Comment by Jongy** [comment]
I know that's how the feature is defined, but at least for my use case it's unexpected and I'd prefer deep merge instead of shallow merge (where a merged key overrides a previously defined one; I actually wanted them to be deep merged). IMO yamllint as a linter is here to assist where the behavior defined by the standard can get users get be confused.

Anyway I hacked through the code shortly and realized it's not an easy thing to add since yamllint has no notion of anchors at the moment :/ (an undefined anchor `anch` expanded with `*anch` doesn't fail `yamllint --strict ...`). Once yamllint keeps tracks of defined anchors, it'll be fairly easy to check [here](https://github.com/adrienverge/yamllint/blob/e8391de71173484c2ae691827aebae62752ce9ca/yamllint/rules/key_duplicates.py#L93) if the key is also present in any of the merged anchors (recursively)

**Comment by MarkusTeufelberger** [comment]
It is irrelevant that YAML has a defined behavior for this. Lots of syntax that is technically valid YAML is seen as problematic by this linter, there should be at least a warning that there is a collision when merging because I can guarantee you that even though this is defined a lot of people will not know the definition.

**Comment by adrienverge** [comment]
```yaml
.foo: &foo
  my_key: value1

bar:
  <<: *foo
  my_key: value2
```
is actually processed into:
```yaml
.foo:
  my_key: value1
bar:
  my_key: value2
```

Like @perlpunk said, the whole point of anchors and aliases is to be able to merge objects, and redefine/override keys. I understand that you would prefer "deep merge" to "replacement merge", but this is not how YAML works. This topic isn't related to yamllint I think.

We could imagine a new option or new rule in yamllint to *forbid overriding keys in aliases if these keys have already been declared in the related anchor*, but it won't be an easy task. I'm not sure there is a need for this niche problem (which in my opinion results from a surprise when discovering YAML, and isn't a problem at all once you understand what anchors/aliases are for).

**Comment by MarkusTeufelberger** [comment]
Personally I don't have a preference between deep or replacement merges, I definitely would start looking into either annotating source or destination of my merges if yamllint told me that there are values being overwritten in a merge though.

According to the description:

> yamllint does not only check for syntax validity, but for weirdnesses like key repetition and cosmetic problems such as lines length, trailing spaces, indentation, etc.

Octal values for example are also perfectly fine and within spec, yet there is a rule in place in yamllint to warn you about potential unexpected consequences.

I would expect to get a warning about a merge actually overriding some data structure(s) (ideally including some hints which data structure(s) exactly are affected) as this is likely to cause some pain down the road due to wrong or no expectations towards what the outcome of such a merge would be. Just like `010 != 10` which is a "weirdness" in the spec, I consider this merging behavior (no matter which way it is defined) a "weirdness" too that should at least be brought to people's attention. It would be easy to disable that rule on a case-by-case basis if it is something that's heavily used in some environments (build definitions?), but I expect some big aha moments in many places once something like this at least creates a warning message with an explanation that this is in the spec but might be unintended.

**Comment by adrienverge** [comment]
I get your point, but:

> Personally I don't have a preference between deep or replacement merges, I definitely would start looking into either annotating source or destination of my merges if yamllint told me that there are values being overwritten in a merge though.

Even if you create a new rule for this purpose, yamllint won't "*tell you that there are values being overwritten in a merge, so you can start looking if something's wrong*", unless you proactively enable that special rule in yamllint configuration. Such a new rule cannot be enabled by default: these values being overwritten is the normal behavior, that's how YAML works, and that's what most users expect (because it's pretty useful).
